### PR TITLE
Bleu memory leak

### PR DESCRIPTION
**Pull request by artemru**
**What does this PR do? Please describe:**
A summary of the change or the issue that is fixed.

Fixes #{issue number}

**Does your PR introduce any breaking changes? If yes, please list them:**
List of all backwards-incompatible changes.

**Check list:**
- [ ] Was the content of this PR **discussed and approved** via a GitHub issue? (no need for typos or documentation improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairseq2/blob/main/CONTRIBUTING.md)?
- [ ] Did you make sure that your **PR does only one thing** instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**?
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you **update the [CHANGELOG](https://github.com/facebookresearch/fairseq2/blob/main/CHANGELOG.md)**? (no need for typos, documentation, or minor internal changes)

### CODE REVIEW

File: src/fairseq2/metrics/text/_bleu.py [hunk 1]
```diff
@@ -65,14 +76,16 @@ def update(self, refs: Sequence[str], hyps: Sequence[str]) -> Self:
...
         self.sys_len += bleu.sys_len
         self.ref_len += bleu.ref_len
 
-        self.valid_ngrams += torch.tensor(bleu.counts, device=self.device)
-        self.total_ngrams += torch.tensor(bleu.totals, device=self.device)
-
+        self.valid_ngrams += torch.tensor(bleu.counts, device=self.device).clone()
+        self.total_ngrams += torch.tensor(bleu.totals, device=self.device).clone()
+        self._steps += 1
+        if self._steps % 20 == 0:
```
Comment (artemru) [inline src/fairseq2/metrics/text/_bleu.py]: 20 here is arbitrary (too often will create too much time overhead)

Comment (cbalioglu) [reply src/fairseq2/metrics/text/_bleu.py]: Why do we still need to call GC if we have fixed the memory leak? Having said that I think 20 is still too low. In the main training loop, by default ,we call GC every 1000 step (which is configurable via recipe). I think anything less than ~400-500 might start to affect the runtime performance depending on the memory allocation pattern of the recipe

File: src/fairseq2/metrics/text/_bleu.py [hunk 2]
```diff
@@ -65,14 +76,16 @@ def update(self, refs: Sequence[str], hyps: Sequence[str]) -> Self:
...
-        bleu = corpus_bleu(hyps, [refs], tokenize=self.tokenizer)
+        bleu = self._metric_class.corpus_score(hyps, [refs])
 
         self.sys_len += bleu.sys_len
         self.ref_len += bleu.ref_len
 
-        self.valid_ngrams += torch.tensor(bleu.counts, device=self.device)
-        self.total_ngrams += torch.tensor(bleu.totals, device=self.device)
-
+        self.valid_ngrams += torch.tensor(bleu.counts, device=self.device).clone()
```
Comment (cbalioglu) [inline src/fairseq2/metrics/text/_bleu.py]: Why do we clone here an already newly initialized tensor? Seems redundant?

File: src/fairseq2/metrics/text/_bleu.py [hunk 3]
```diff
@@ -57,6 +59,15 @@ def __init__(
 
         self._add_state("valid_ngrams", valid_ngrams)
         self._add_state("total_ngrams", total_ngrams)
+        self._metric_class = BLEU(
```
Comment (artemru) [inline src/fairseq2/metrics/text/_bleu.py:59]: pls, check if I got all the params right

File: src/fairseq2/metrics/text/_bleu.py [hunk 4]
```diff
@@ -16,6 +17,7 @@
 from torcheval.metrics import Metric
 from typing_extensions import Self, override
 
+from fairseq2.logging import log
```
Comment (artemru) [inline src/fairseq2/metrics/text/_bleu.py]: rm

File: src/fairseq2/metrics/text/_bleu.py [hunk 5]
```diff
@@ -65,14 +76,16 @@ def update(self, refs: Sequence[str], hyps: Sequence[str]) -> Self:
...
+        bleu = self._metric_class.corpus_score(hyps, [refs])
 
         self.sys_len += bleu.sys_len
         self.ref_len += bleu.ref_len
 
-        self.valid_ngrams += torch.tensor(bleu.counts, device=self.device)
-        self.total_ngrams += torch.tensor(bleu.totals, device=self.device)
-
+        self.valid_ngrams += torch.tensor(bleu.counts, device=self.device).clone()
+        self.total_ngrams += torch.tensor(bleu.totals, device=self.device).clone()
```
Comment (artemru) [inline src/fairseq2/metrics/text/_bleu.py]: no needed

File: src/fairseq2/metrics/text/_bleu.py [hunk 6]
```diff
@@ -65,14 +76,16 @@ def update(self, refs: Sequence[str], hyps: Sequence[str]) -> Self:
...
+        self._steps += 1
+        if self._steps % 500 == 0:
```
Comment (cbalioglu) [inline src/fairseq2/metrics/text/_bleu.py]: Why do we still need to call GC here if we have fixed the memory leak? Having said that, we call GC every 1000 step in the trainer (which is configurable here: https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/recipes/config.py#L171-L172)

I think a metric implementation should not have this kind of "hidden" behavior that might impact the overall job throughput. Depending on the memory allocation pattern of the job, a GC can become quite expensive.

Comment (artemru) [reply src/fairseq2/metrics/text/_bleu.py]: I can try without it and see if it helps or not!
do we have some gc call after each eval run ?

File: src/fairseq2/metrics/text/_bleu.py (hunk 2 above)
Comment (artemru) [reply src/fairseq2/metrics/text/_bleu.py]: already removed

File: src/fairseq2/metrics/text/_bleu.py (hunk 6 above)
Comment (artemru) [reply src/fairseq2/metrics/text/_bleu.py]: it runs ok without it

Comment (artemru) [reply src/fairseq2/metrics/text/_bleu.py]: @cbalioglu : does `gc_every_n_steps` acts also for validation steps ?

Comment (artemru) [reply src/fairseq2/metrics/text/_bleu.py]: actually, no - it's better with an internal gc.collect() ...

Comment (cbalioglu) [reply src/fairseq2/metrics/text/_bleu.py]: We do a CUDA cache allocator clean-up, but not GC: https://github.com/facebookresearch/fairseq2/blob/main/src/fairseq2/recipes/_trainer.py#L1167-L1168

We can also do GC there though.

> actually, no - it's better with an internal gc.collect()

Do you mean you still get a leak when you don't call gc.collect() even after the refactoring of the corpus call?

File: src/fairseq2/metrics/text/_bleu.py (hunk 3 above)
Comment (gilkeren1) [reply src/fairseq2/metrics/text/_bleu.py:59]: hmmm I personally can't really say
